core/storage/trie/polkadot_trie/polkadot_trie_impl.cpp

```diff
@@ -276,176 +284,181 @@ namespace kagome::storage::trie {
-    // turn branch node left with no children to a leaf node
-    if (bitmap == 0 and parent->value) {
-      newRoot = std::make_shared<LeafNode>(key_nibbles.subspan(0, length),
-                                           parent->value);
-    } else if (parent->childrenNum() == 1 && !parent->value) {
+  outcome::result<void> handleDeletion(
+      PolkadotTrie::NodePtr &parent,
+      const PolkadotTrieImpl::ChildRetrieveFunctor &retrieveChild) {
+    auto branch = std::dynamic_pointer_cast<BranchNode>(parent);
+    auto bitmap = branch->childrenBitmap();
+    if (bitmap == 0) {
+      if (parent->value) {
+        // turn branch node left with no children to a leaf node
+        parent = std::make_shared<LeafNode>(parent->key_nibbles, parent->value);
+      } else {
+        // this case actual only for clearPrefix, unreal situation in deletion
+        parent.reset();
+      }
+    } else if (branch->childrenNum() == 1 && !parent->value) {
       size_t idx = 0;
       for (idx = 0; idx < 16; idx++) {
         bitmap >>= 1u;
         if (bitmap == 0) {
           break;
         }
       }
-      OUTCOME_TRY(child, retrieveChild(parent, idx));
+      OUTCOME_TRY(child, retrieveChild(branch, idx));
       using T = PolkadotNode::Type;
       if (child->getTrieType() == T::Leaf) {
-        auto newKey = parent->key_nibbles;
-        newKey.putUint8(idx);
-        newKey.putBuffer(child->key_nibbles);
-        newRoot = std::make_shared<LeafNode>(newKey, child->value);
-      } else if (child->getTrieType() == T::BranchEmptyValue
-                 or child->getTrieType() == T::BranchWithValue) {
-        auto branch = std::make_shared<BranchNode>();
-        branch->key_nibbles.putBuffer(parent->key_nibbles)
-            .putUint8(idx)
-            .putBuffer(child->key_nibbles);
-        auto child_as_branch = std::dynamic_pointer_cast<BranchNode>(child);
-        for (size_t i = 0; i < child_as_branch->children.size(); i++) {
-          if (child_as_branch->children.at(i)) {
-            branch->children.at(i) = child_as_branch->children.at(i);
-          }
-        }
-        branch->value = child->value;
-        newRoot = branch;
+        parent = std::make_shared<LeafNode>(parent->key_nibbles, child->value);
+      } else if (child->isBranch()) {
+        branch->children = dynamic_cast<BranchNode &>(*child.get()).children;
+        parent->value = child->value;
       }
+      parent->key_nibbles.putUint8(idx).putBuffer(child->key_nibbles);
     }
-    return newRoot;
+    return outcome::success();
   }
 
-  outcome::result<std::tuple<PolkadotTrie::NodePtr, size_t>>
-  PolkadotTrieImpl::detachNode(const NodePtr &parent,
-                               const KeyNibbles &prefix_nibbles,
-                               const OnDetachCallback &callback) {
-    size_t count = 0;
+  outcome::result<void> notifyOnDetached(
+      PolkadotTrie::NodePtr &, const PolkadotTrie::OnDetachCallback &);
+
+  outcome::result<void> detachNode(
```

Comment:
There are 7 parameters and 0 lines of doc string. Moreover, half of them are passed as non-const refs. Please add the docstring. Otherwise, we have to guess what the `count` value stands for without exploring the method sources.